### DISCUSSION TITLE
videodownloader help code error

### DISCUSSION THREAD

**Discussion by 77tpro11** [aclap-dev/video-downloadhelper#588] [Help and Q&A]
### Your question

{
  "version": "9.0.1.13",
  "target": "google",
  "channel": "stable",
  "page_url": "https://www.pornhub.com/view_video.php?viewkey=ph6177056997d3f",
  "error": {
    "details": "SideDownload error: : [tcp @ 000001d15f53c580] Connection to tcp://ev-h.phncdn.com:443 failed: Error number -138 occurred\r\n[in#0 @ 000001d15d2c8740] Error opening input: Error number -138 occurred\r\nError opening input file https://ev-h.phncdn.com/hls/videos/202110/25/396965911/,720P_4000K,480P_2000K,240P_1000K,_396965911.mp4.urlset/index-f1-v1-a1.m3u8?validfrom=1716498309&validto=1716505509&ipa=172.58.229.250&hdl=-1&hash=2xfUUU0xDGQWnpTTZVPmogiXeX4%3D.\r\nError opening input files: Error number -138 occurred\r\n",
    "downloadable_id": "downloadable_1203057555051226",
    "error": "coapp_failure",
    "id": "coapp_failure_6dbefcef-622c-44b5-926a-4661b356b800"
  }
}

### Addon & CoApp version and details

{
  "version": "9.0.1.13",
  "target": "google",
  "channel": "stable",
  "page_url": "https://www.pornhub.com/view_video.php?viewkey=ph6177056997d3f",
  "error": {
    "details": "SideDownload error: : [tcp @ 000001d15f53c580] Connection to tcp://ev-h.phncdn.com:443 failed: Error number -138 occurred\r\n[in#0 @ 000001d15d2c8740] Error opening input: Error number -138 occurred\r\nError opening input file https://ev-h.phncdn.com/hls/videos/202110/25/396965911/,720P_4000K,480P_2000K,240P_1000K,_396965911.mp4.urlset/index-f1-v1-a1.m3u8?validfrom=1716498309&validto=1716505509&ipa=172.58.229.250&hdl=-1&hash=2xfUUU0xDGQWnpTTZVPmogiXeX4%3D.\r\nError opening input files: Error number -138 occurred\r\n",
    "downloadable_id": "downloadable_1203057555051226",
    "error": "coapp_failure",
    "id": "coapp_failure_6dbefcef-622c-44b5-926a-4661b356b800"
  }
}

**Comment by paulrouget** [comment]
ff-failure: 403. I'll do some testing.

**Comment by GrampaWildWilly** [comment]
This is another one of those that if you reload the page enough times, VDH will manage to download the content.  I did manage to eventually download this, after several page reloads.  This is a problem we have encountered before with Pornhub.  It seems to depend on the direction the wind is blowing, the phase of the moon, how long it's been since the last time you brushed you teeth, & several other equally logical factors.

I also tried to download this with ffmpeg directly.  I never did manage to succeed in doing it.  First off, there's Timed_ID3 data being foisted on us here.  I assume that since VDH uses ffmpeg now, that this is no longer the problem it used to be.  Ffmpeg doesn't know what to do with Timed_ID3 data so it just ignores it, which is the right thing to do.

I also kept seeing 403 errors but it seemed like they were 403 errors on particular chunks of the stream.  It would read a certain amount of the content & then get stuck with 403 errors.  Ffmpeg never managed to download more than a 50-second section of this video, & that didn't seem to be coming from the start of it.

On many of my attempts to get this with ffmpeg, it would say certain streams were corrupt.  The HLS master manifests always said there were 4 Programs of various resolutions.  The 1280x720 variant seemed to always be recognized correctly.  But the 3 lower resolution variants would be recognized properly sometimes & not other times.

Another anomalous thing I noticed was the duration of the scene was being reported as N/A by ffmpeg, as if it were a livestream.  So there were multiple ways the site was not reporting things correctly.

I tried another video chosen at random elsewhere on Pornhub.  It took me 3 or 4 page reloads & VDH did successfully download that.

I believe that if we wait a few hours, things will start working reliably on the site.  Over the years, I have downloaded stuff off there without any problems using previous versions of VDH.  I think there are just times when the site is acting up & you just have to be patient.  Wait until it goes back to normal.

**Reply by paulrouget** [reply]
Could be a cache issue. Maybe force reload the page solve it?

I'm thinking of adding a "Force Reload Page" button in the "No Media" box and in the "Download Fail" box.

**Reply by GrampaWildWilly** [reply]
I was doing Ctrl+F5 page reloads.  Didn't seem to matter.

**Comment by GrampaWildWilly** [comment]
I decided to try this again but with a slight modification.  I mentioned that the HLS master manifest was reporting a duration of N/A as if it were a livestream.  So I decided to take it at its word.  There are 3 parameters I use with ffmpeg only when it's a livestream.  -m3u8_hold_counters 25 & -max_reload 25 control the way ffmpeg detects the end of a livestream.  If you don't code these parameters, you can get ffmpeg reading the manifest 1,000 times before it decides the manifest is not going to get updated & it terminates the download.  My values scale that back to just 25 times.  -live_start_index 0 rewinds the livestream to the beginning.  This requires that the server actually stores the stream from the beginning.  If not, this parameter will rewind as far as the server will allow.  But in this case, it is NOT a livestream.  So my 2 parameters for controlling the termination didn't come into play.  But the rewind parameter did cause the recording to start at the beginning of the clip.

Everything was running smoothly until it got to the same chunk.  You can tell which chunk it is because ffmpeg logs the URLs of the chunks as it reads them.  On most web sites, these URLs include a number somewhere buried in amongst all the gibberish characters that increments by 1 with each chunk.  Fortunately, Pornhub at least follows this convention.  So after reading about 4 1/2 minutes of this, it got to that same old problematic chunk & started getting repeated 403 errors.  Ffmpeg eventually decided it had seen enough errors & terminated the download.  I did have the first 4 1/2 minutes of this, which is better that the maximum 50 seconds I was getting earlier.  And it was from the beginning.  But it still wasn't the whole video.

I don't know what's up with this.  I'm still convinced that if we wait long enough, both VDH & ffmpeg will get this video no problem, all of it, first try, every try.  I'm still convinced this is not a VDH problem, that it's a problem at the Pornhub server(s).

**Reply by GrampaWildWilly** [reply]
Not that I'm bragging or anything, but my prediction was correct.  I tried the video in the original post above again just now.  I downloaded it just fine on the first try with VDH.  No problems.  I also downloaded it just fine at the same time using ffmpeg directly.  That also worked just fine, first try no problems.

Then I reloaded VDH & Ctrl+F5 reloaded the web page, then did both downloads again.  First try.  No problems.

Well, there was one problem, but it's something we've observed before.  Both VDH & direct ffmpeg downloads stopped at a certain point.  They didn't fail, they just paused for maybe 2 or 3 minutes.  The stoppage was mysterious, & then they both mysteriously resumed.  Then they completed normally, as I already said.  These mysterious stoppages appear to me as the Pornhub server just deciding to take a bathroom break.  Then it comes back & things resume like everything was fine, nothing happened, whatchoo bitchinbout?  I don't know if you have any better insights on these breaks, Paul.  I think there's nothing that can be done about it.  We just have to be patient & wait it out.  As long as you don't see an outright error and/or termination, just have faith & let it run.

Oh.  I see that I have failed to mention this.  I'm running Windows 11 Pro 64-bit version 23H2 build 22631.3593, Firefox 126.0 64-bit, licensed VDH 9.0.1.22a1 beta, CoApp 2.0.19.
